### PR TITLE
Improve --optimize-images to find images nested within XObjects

### PR DESCRIPTION
**Pull request by cdosborn**
See this issue: https://github.com/qpdf/qpdf/issues/923

@jberkenbilt Appreciate any feedback you have! How would you suggest me testing this?

### CODE REVIEW

**Comment by jberkenbilt** [comment]
We'll have to update the test suite with a file that has this. You can search in qpdf/qtest/*.test for where `--optimize-images` is tested and see about adding a new test file. The test file has to be freely redistributed, unlicensed, etc.

**Comment by cdosborn** [comment]
Getting the following diff when I run my test locally:
```
-qpdf: image /X1 on page 1: optimizing image reduces size from 2628 to 1592
+qpdf: image /X1 on page 1: optimizing image reduces size from 2628 to ...
 qpdf: wrote file a.pdf
```
Trying to figure out why this number is being truncated via ellipsis. Seems to be an issue with logging?

**Comment by cdosborn** [comment]
> Getting the following diff when I run my test locally:
> 
> ```
> -qpdf: image /X1 on page 1: optimizing image reduces size from 2628 to 1592
> +qpdf: image /X1 on page 1: optimizing image reduces size from 2628 to ...
>  qpdf: wrote file a.pdf
> ```
> 
> Trying to figure out why this number is being truncated via ellipsis. Seems to be an issue with logging?

Figured it out.

File: libqpdf/QPDFJob.cc [hunk 1]
```diff
@@ -2350,30 +2350,34 @@ QPDFJob::handleTransformations(QPDF& pdf)
...
+                        image);
+                    std::shared_ptr<QPDFObjectHandle::StreamDataProvider> sdp(
+                        io);
+                    if (io->evaluate(
+                            "image " + name + " on page " +
+                            std::to_string(pageno))) {
+                        QPDFObjectHandle new_image = pdf.newStream();
+                        new_image.replaceDict(image.getDict().shallowCopy());
+                        new_image.replaceStreamData(
+                            sdp,
```
Comment (m-holger) [inline libqpdf/QPDFJob.cc:2375]: ... and
```cpp
                                std::move(io),
```
it here.

File: libqpdf/QPDFJob.cc [hunk 2]
```diff
@@ -2350,30 +2350,34 @@ QPDFJob::handleTransformations(QPDF& pdf)
...
+                    std::string name = key;
+                    QPDFObjectHandle& image = obj;
+                    ImageOptimizer* io = new ImageOptimizer(
+                        *this,
+                        m->oi_min_width,
+                        m->oi_min_height,
+                        m->oi_min_area,
+                        image);
+                    std::shared_ptr<QPDFObjectHandle::StreamDataProvider> sdp(
+                        io);
```
Comment (m-holger) [inline libqpdf/QPDFJob.cc:2368]: @jberkenbilt  It probably would be cleaner to create `io` as a unique_ptr 
```cpp
                    auto io = std::make_unique<ImageOptimizer>(
                        *this,
                        m->oi_min_width,
                        m->oi_min_height,
                        m->oi_min_area,
                        image);
```
and then ...

File: libqpdf/QPDFJob.cc (hunk 1 above)
Comment (cdosborn) [reply libqpdf/QPDFJob.cc:2375]: Feel free to push to my branch. I'll need to spend some time to understand this.

File: libqpdf/QPDFJob.cc [hunk 3]
```diff
@@ -2350,30 +2350,33 @@ QPDFJob::handleTransformations(QPDF& pdf)
...
-                        .replaceKey(name, new_image);
-                }
-            }
+            ph.forEachImage(
+                true,
+                [this, pageno, &pdf](
+                    QPDFObjectHandle& obj,
+                    QPDFObjectHandle& xobj_dict,
+                    std::string const& key) {
+                    QPDFObjectHandle& image = obj;
```
Comment (m-holger) [inline libqpdf/QPDFJob.cc:2359]: Sorry, also no need to create a new to create a new ref to obj.

(As an aside, `(auto& image, auto& xobj_dict, auto const& key)` also works.)

File: libqpdf/QPDFJob.cc [hunk 4]
```diff
@@ -2350,30 +2350,33 @@ QPDFJob::handleTransformations(QPDF& pdf)
         for (auto& ph: dh.getAllPages()) {
             ++pageno;
             QPDFObjectHandle page = ph.getObjectHandle();
```
Comment (m-holger) [inline libqpdf/QPDFJob.cc:2352]: I think `page` is no longer used

File: libqpdf/QPDFJob.cc [hunk 5]
```diff
@@ -2350,30 +2350,34 @@ QPDFJob::handleTransformations(QPDF& pdf)
...
-                        .replaceKey(name, new_image);
-                }
-            }
+            QPDFPageObjectHelper(page).forEachImage(
+                true,
+                [&, pageno, &pdf = pdf](
+                    QPDFObjectHandle& obj,
+                    QPDFObjectHandle& xobj_dict,
+                    std::string const& key) mutable {
+                    std::string name = key;
```
[the remark this answers is not in the document]
Comment (cdosborn) [reply libqpdf/QPDFJob.cc]: :sweat_smile: Didn't realize that this constructed a new object.

**Comment by cdosborn** [comment]
@m-holger Thanks for feedback. Let me know if there is anything else I can improve.

**Review by jberkenbilt — APPROVED** [review]
Thanks @cdosborn for the contribution and @m-holger for the review. All I have to do is click the button! (But I also reviewed -- looks great.)

**Pull request merged**